### PR TITLE
refactor(payment_methods): add a new domain type for payment method data to be used in connector module

### PR DESCRIPTION
**Pull request by Narayanbhat166**
## Type of Change

- [x] Refactoring

## Description

This PR adds a new domain type for `payment_method_data` to be used everywhere in the application.

## Motivation and Context

We need to have separate types for api model and domain, currently we are using the api type in application as well. 

## How did you test it?

- Sanity test of payments.

# Impact
- No impact on any flow or connector

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`

### CODE REVIEW

File: crates/router/src/core/payments.rs [hunk 1]
```diff
@@ -2207,6 +2207,89 @@ pub mod payment_address {
...
+
+            Self {
+                shipping: self.shipping,
+                billing: self.billing,
+                unified_payment_method_billing,
+                payment_method_billing: self.payment_method_billing,
+            }
+        }
+
+        /// Merge / Unify details from `payment.payment_method_data.[payment_method_data]` with the billig address.
```
Comment (ArjunKarthik) [inline crates/router/src/core/payments.rs:2228]:
```suggestion
        /// Merge / Unify details from `payment.payment_method_data.[payment_method_data]` with the billing address.
        /// The unification here takes place on the basis of leaf node values
```

**Review by ArjunKarthik — DISMISSED** [review]

File: crates/api_models/src/payments.rs [hunk 2]
```diff
@@ -2639,6 +2642,33 @@ pub struct AddressDetails {
     pub last_name: Option<Secret<String>>,
 }
 
+impl AddressDetails {
+    pub fn unify_address_details(self, other: Option<&Self>) -> Self {
+        match (self, other) {
```
Comment (hrithikesh026) [inline crates/api_models/src/payments.rs:2647]: matching self is not required i guess.

File: crates/router/build.rs [hunk 3]
```diff
@@ -2,7 +2,7 @@ fn main() {
     // Set thread stack size to 4 MiB for debug builds
     // Reference: https://doc.rust-lang.org/std/thread/#stack-size
     #[cfg(debug_assertions)]
-    println!("cargo:rustc-env=RUST_MIN_STACK=6291456"); // 6 * 1024 * 1024 = 6 MiB
+    println!("cargo:rustc-env=RUST_MIN_STACK=8291456"); // 6 * 1024 * 1024 = 6 MiB
```
Comment (hrithikesh026) [inline crates/router/build.rs:5]: Please update the comment as well

**Review by ThisIsMani — DISMISSED** [review]
Dashboard specific changes looks fine.

**Review by hrithikesh026 — APPROVED** [review]
LGTM

**Review by ArjunKarthik — DISMISSED** [review]

**Review by hrithikesh026 — APPROVED** [review]

**Review by ThisIsMani — APPROVED** [review]

**Review by ArjunKarthik — APPROVED** [review]

**Pull request merged**